File: api/src/test/java/net/blueshell/api/validation/AddressValidatorTest.java
package net.blueshell.api.validation;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validator;
import net.blueshell.api.dto.AddressDTO;
import net.blueshell.api.factory.dto.AddressDTOFactory;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Unit tests for AddressDTO validation.
 */
@SpringBootTest
class AddressValidatorTest {

    private final Validator validator;
    private final AddressDTOFactory addressFactory;

    @Autowired
    AddressValidatorTest(Validator validator, AddressDTOFactory addressFactory) {
        this.validator = validator;
        this.addressFactory = addressFactory;
    }

    @Test
    void validAddressDTO_passesValidation() {
        AddressDTO dto = addressFactory.createBasic();
        Set<ConstraintViolation<AddressDTO>> violations = validator.validate(dto);
        assertTrue(violations.isEmpty(), "Valid AddressDTO should pass validation");
    }

    @Test
    void addressDTO_withInvalidCountryCode_failsValidation() {
        AddressDTO dto = addressFactory.createWithCustomizations(a -> a.setCountry("INVALID"));
        Set<ConstraintViolation<AddressDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("country")));
    }

    @Test
    void addressDTO_withEmptyCity_failsValidation() {
        AddressDTO dto = addressFactory.createWithCustomizations(a -> a.setCity(""));
        Set<ConstraintViolation<AddressDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("city")));
    }

    @Test
    void addressDTO_withEmptyStreet_failsValidation() {
        AddressDTO dto = addressFactory.createWithCustomizations(a -> a.setStreet(""));
        Set<ConstraintViolation<AddressDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("street")));
    }

    @Test
    void addressDTO_withEmptyHouseNumber_failsValidation() {
        AddressDTO dto = addressFactory.createWithCustomizations(a -> a.setHouseNumber(""));
        Set<ConstraintViolation<AddressDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("houseNumber")));
    }

    @Test
    void addressDTO_withEmptyZipCode_failsValidation() {
        AddressDTO dto = addressFactory.createWithCustomizations(a -> a.setZipCode(""));
        Set<ConstraintViolation<AddressDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("zipCode")));
    }
}

File: api/src/test/kotlin/net/blueshell/api/domain/user/web/validation/AddressValidatorTest.kt
package net.blueshell.api.domain.user.web.validation

import jakarta.validation.ConstraintViolation
import jakarta.validation.Validator
import net.blueshell.api.factory.dto.AddressDTOFactory
import net.blueshell.api.domain.user.web.dto.AddressDTO
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

/**
 * Unit tests for AddressDTO validation.
 */
@SpringBootTest
class AddressValidatorTest @Autowired constructor(
    private val validator: Validator,
    private val addressFactory: AddressDTOFactory
) {

    @Test
    fun `valid address dto passes validation`() {
        val dto = addressFactory.createBasic()
        val violations: Set<ConstraintViolation<AddressDTO>> = validator.validate(dto)
        assertTrue(violations.isEmpty(), "Valid AddressDTO should pass validation")
    }

    @Test
    fun `address dto with invalid country code fails validation`() {
        val dto = addressFactory.createWithCustomizations { it.country = "INVALID" }
        val violations: Set<ConstraintViolation<AddressDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "country" })
    }

    @Test
    fun `address dto with empty city fails validation`() {
        val dto = addressFactory.createWithCustomizations { it.city = "" }
        val violations: Set<ConstraintViolation<AddressDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "city" })
    }

    @Test
    fun `address dto with empty street fails validation`() {
        val dto = addressFactory.createWithCustomizations { it.street = "" }
        val violations: Set<ConstraintViolation<AddressDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "street" })
    }

    @Test
    fun `address dto with empty house number fails validation`() {
        val dto = addressFactory.createWithCustomizations { it.houseNumber = "" }
        val violations: Set<ConstraintViolation<AddressDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "houseNumber" })
    }

    @Test
    fun `address dto with empty zip code fails validation`() {
        val dto = addressFactory.createWithCustomizations { it.zipCode = "" }
        val violations: Set<ConstraintViolation<AddressDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "zipCode" })
    }
}
